layerindex/tools/import_layers.py

```diff
@@ -0,0 +1,494 @@
+    return jsdict
+
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Layer index import utility. Imports layer information from another layer index instance using the REST API. WARNING: this will overwrite data in your database, use with caution!")
+    parser.add_argument('url', help='Layer index URL to fetch from')
+    parser.add_argument('-b', '--branch', action='store', help='Restrict to import a specific branch only (separate multiple branches with commas)')
+    parser.add_argument('-l', '--layer', action='store', help='Restrict to import a specific layer only (regular expressions allowed)')
+    parser.add_argument('-r', '--reload', action='store_true', help='Reload data even if it is up-to-date')
+    parser.add_argument('-n', '--dry-run', action='store_true', help="Don't write any data back to the database")
+    parser.add_argument('-d', '--debug', action='store_true', help='Enable debug output')
+    parser.add_argument('-q', '--quiet', action='store_true', help='Hide all output except error messages')
+
+    args = parser.parse_args()
+
+    if args.debug:
+        loglevel = logging.DEBUG
+    elif args.quiet:
+        loglevel = logging.WARNING
+    else:
+        loglevel = logging.INFO
+
+    utils.setup_django()
+    import settings
+    from layerindex.models import Branch, LayerItem, LayerBranch, LayerDependency, LayerMaintainer, LayerNote, Recipe, Source, Patch, PackageConfig, StaticBuildDep, DynamicBuildDep, RecipeFileDependency, Machine, Distro, BBClass, BBAppend, IncFile
+    from django.db import transaction
+
+    logger.setLevel(loglevel)
+
+    fetchdir = settings.LAYER_FETCH_DIR
+    if not fetchdir:
+        logger.error("Please set LAYER_FETCH_DIR in settings.py")
+        sys.exit(1)
+
+    if not os.path.exists(fetchdir):
+        os.makedirs(fetchdir)
+
+    layerindex_url = args.url
+    if not layerindex_url.endswith('/'):
+        layerindex_url += '/'
+    if not '/layerindex/api/' in layerindex_url:
+        layerindex_url += 'layerindex/api/'
+
+    rq = urllib.request.Request(layerindex_url)
+    data = urllib.request.urlopen(rq).read()
+    jsdata = json.loads(data.decode('utf-8'))
+
+    branches_url = jsdata['branches']
+    layers_url = jsdata['layerItems']
+    layerdeps_url = jsdata['layerDependencies']
+    layerbranches_url = jsdata['layerBranches']
+    layermaintainers_url = jsdata.get('layerMaintainers', None)
+    layernotes_url = jsdata.get('layerNotes', None)
+    recipes_url = jsdata.get('recipesExtended', None)
+    machines_url = jsdata.get('machines', None)
+    distros_url = jsdata.get('distros', None)
+    classes_url = jsdata.get('classes', None)
+    appends_url = jsdata.get('appends', None)
+    incfiles_url = jsdata.get('incFiles', None)
+
+    logger.debug('Getting branches')
+
+    # Get branches (we assume the ones we want are already there, so skip any that aren't)
+    rq = urllib.request.Request(branches_url)
+    data = urllib.request.urlopen(rq).read()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *urlopen*. The resource is allocated by call *request.urlopen*. Execution paths that do not contain closure statements were detected. To prevent this resource leak, close the object returned by *urlopen()* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).